### PR TITLE
Return a reference with all entitlements, when access through owned values

### LINKED ISSUE #2536: Improve entitlement mappings
### Issue to be solved

This is to track the implementation of https://github.com/onflow/flips/pull/94

### Suggested Solution

_No response_

### PR DESCRIPTION
**Pull request by SupunS**
Work towards #2638
Work towards onflow/flips#94
Closes #2536

## Description

Adds the improvement discussed in https://github.com/onflow/cadence/pull/2588#discussion_r1237203195

______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/sema/check_member_expression.go
Comment (SupunS) [inline runtime/sema/check_member_expression.go:378]: This function didn't need to be a closure, so extracted it to a top-level function

**Review by turbolent — APPROVED** [review]
Nice!

**Comment by turbolent** [comment]
It would be good to get a review from @dsainati1 here before merging

File: runtime/sema/check_member_expression.go
Comment (dsainati1) [inline runtime/sema/check_member_expression.go:473]: This is used for accessor functions right? That use case makes sense, but we should add a test to make sure that a function-typed field with an identity mapping should not get the entitlements to return type. I.e. some field defined with `access(Identity) let fn: ((): T)` should not be entitled to `T` when it's accessed through an owned value.

File: runtime/interpreter/interpreter.go
Comment (SupunS) [reply runtime/interpreter/interpreter.go:5241]: Can you add this requirement to the feature-request https://github.com/onflow/cadence/issues/2643.

(I guess there's nothing to add/change as for now / this PR?)

File: runtime/tests/checker/entitlements_test.go
Comment (dsainati1) [reply runtime/tests/checker/entitlements_test.go:5479]: Oh I just realized why this doesn't make sense; if the value is owned we won't produce a reference. Nevermind this 😄

File: runtime/sema/check_member_expression.go
Comment (SupunS) [reply runtime/sema/check_member_expression.go:473]: Didn't quite understand the question. For an owned value, `fn` won't return a reference, so we are not actually computing any entitlements for reference, the field is readable for anyone since the value is owned, correct?

On the other hand, it is not allowed to use Mappings on function-typed fields. Added a test e2ccbb0edf51eee1f7af6a8f520fa31b4875252d.

File: runtime/tests/checker/entitlements_test.go
Comment (dsainati1) [inline runtime/tests/checker/entitlements_test.go:5555]: Oh hmm, if functions are just disallowed here maybe use a reference to a function type so we can actually test the mapping behavior.

File: runtime/sema/check_member_expression.go
Comment (dsainati1) [reply runtime/sema/check_member_expression.go:473]: Sorry, I should have been clearer. In https://github.com/onflow/cadence/pull/2637#discussion_r1258458218 I suggested adding a field with a `&((): T)` type to exercise this case.

Comment (SupunS) [reply runtime/sema/check_member_expression.go:473]: I see what you mean.

Weird enough, in stable cadence (not just this feature branch), accessing a field of type `&((): T)`, returns a `((): T)`. Don't know if this was changed intentionally, or was a regression (to me it seems like a regression, and probably needs fixing). So the use-case you mentioned, is currently not possible. But if we change that other behavior, then yeah, we might run into what you said.

I will add a fix to that hypothetical scenario for now, so it should seamlessly work, once the other issue is fixed.

Good catch nevertheless!

Comment (SupunS) [reply runtime/sema/check_member_expression.go]: added in 842a9cd482a53e47943756ca6817c45774c531ea

File: runtime/tests/checker/entitlements_test.go
Comment (SupunS) [reply runtime/tests/checker/entitlements_test.go:5555]: added in 842a9cd482a53e47943756ca6817c45774c531ea

**Review by dsainati1 — APPROVED** [review]
Looks good to me! Thanks for investigating all the issues here.

**Pull request merged**
